File: backend/src/modules/reports/aggregates.ts
import { and, eq, gte, inArray, lte, sql } from "drizzle-orm";
import { db } from "../../db/index.js";
import {
  accounts,
  journalEntries,
  journalLines,
  openItems,
} from "../../db/schema.js";

function toNum(value: string | null | undefined): number {
  return Number(value ?? 0);
}

export interface DashboardMetrics {
  cash: number;
  periodRevenue: number;
  periodNetIncome: number;
  receivables: number;
  payables: number;
  equity: number;
}

export interface DateRange {
  from?: string;
  to?: string;
}

function defaultPeriod(): { from: string; to: string } {
  const now = new Date();
  const from = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-01`;
  const to = now.toISOString().slice(0, 10);
  return { from, to };
}

async function accountBalances(orgId: string, asOf?: string) {
  const conditions = [
    eq(journalEntries.orgId, orgId),
    eq(journalEntries.status, "posted"),
  ];
  if (asOf) {
    conditions.push(lte(journalEntries.entryDate, asOf));
  }

  const rows = await db
    .select({
      accountId: accounts.id,
      code: accounts.code,
      name: accounts.name,
      type: accounts.type,
      isCash: accounts.isCash,
      debit: sql<string>`coalesce(sum(${journalLines.debit}), 0)`,
      credit: sql<string>`coalesce(sum(${journalLines.credit}), 0)`,
    })
    .from(journalLines)
    .innerJoin(journalEntries, eq(journalLines.entryId, journalEntries.id))
    .innerJoin(accounts, eq(journalLines.accountId, accounts.id))
    .where(and(...conditions))
    .groupBy(
      accounts.id,
      accounts.code,
      accounts.name,
      accounts.type,
      accounts.isCash,
    );

  return rows.map((row) => {
    const debit = toNum(row.debit);
    const credit = toNum(row.credit);
    const balance =
      row.type === "asset" || row.type === "expense"
        ? debit - credit
        : credit - debit;
    return { ...row, debit, credit, balance };
  });
}

async function periodActivity(orgId: string, from: string, to: string) {
  const rows = await db
    .select({
      type: accounts.type,
      debit: sql<string>`coalesce(sum(${journalLines.debit}), 0)`,
      credit: sql<string>`coalesce(sum(${journalLines.credit}), 0)`,
    })
    .from(journalLines)
    .innerJoin(journalEntries, eq(journalLines.entryId, journalEntries.id))
    .innerJoin(accounts, eq(journalLines.accountId, accounts.id))
    .where(
      and(
        eq(journalEntries.orgId, orgId),
        eq(journalEntries.status, "posted"),
        gte(journalEntries.entryDate, from),
        lte(journalEntries.entryDate, to),
      ),
    )
    .groupBy(accounts.type);

  let periodRevenue = 0;
  let periodExpense = 0;

  for (const row of rows) {
    const debit = toNum(row.debit);
    const credit = toNum(row.credit);
    if (row.type === "revenue") periodRevenue += credit - debit;
    if (row.type === "expense") periodExpense += debit - credit;
  }

  return { periodRevenue, periodExpense, periodNetIncome: periodRevenue - periodExpense };
}

export async function getDashboardMetrics(
  orgId: string,
  range?: DateRange,
): Promise<DashboardMetrics> {
  const { from, to } = range?.from && range?.to ? range as { from: string; to: string } : defaultPeriod();
  const balances = await accountBalances(orgId, to);
  const activity = await periodActivity(orgId, from, to);

  const cash = balances
    .filter((a) => a.isCash)
    .reduce((sum, a) => sum + a.balance, 0);

  const receivableItems = await db
    .select({ balance: openItems.balanceAmount })
    .from(openItems)
    .where(and(eq(openItems.orgId, orgId), eq(openItems.type, "receivable")));

  const payableItems = await db
    .select({ balance: openItems.balanceAmount })
    .from(openItems)
    .where(and(eq(openItems.orgId, orgId), eq(openItems.type, "payable")));

  let receivables = receivableItems.reduce((s, i) => s + toNum(i.balance), 0);
  let payables = payableItems.reduce((s, i) => s + toNum(i.balance), 0);

  if (receivables === 0) {
    receivables = balances
      .filter((a) => a.type === "asset" && a.code.startsWith("13"))
      .reduce((s, a) => s + a.balance, 0);
  }
  if (payables === 0) {
    payables = balances
      .filter((a) => a.type === "liability" && a.code.startsWith("21"))
      .reduce((s, a) => s + a.balance, 0);
  }

  const equityAccounts = balances
    .filter((a) => a.type === "equity")
    .reduce((s, a) => s + a.balance, 0);

  const allTimeRevenue = balances
    .filter((a) => a.type === "revenue")
    .reduce((s, a) => s + a.balance, 0);
  const allTimeExpense = balances
    .filter((a) => a.type === "expense")
    .reduce((s, a) => s + a.balance, 0);

  const equity = equityAccounts + (allTimeRevenue - allTimeExpense);

  return {
    cash,
    periodRevenue: activity.periodRevenue,
    periodNetIncome: activity.periodNetIncome,
    receivables,
    payables,
    equity,
  };
}

export async function getProfitLoss(orgId: string, from: string, to: string, breakdown = false) {
  const rows = await db
    .select({
      accountId: accounts.id,
      code: accounts.code,
      name: accounts.name,
      type: accounts.type,
      debit: sql<string>`coalesce(sum(${journalLines.debit}), 0)`,
      credit: sql<string>`coalesce(sum(${journalLines.credit}), 0)`,
    })
    .from(journalLines)
    .innerJoin(journalEntries, eq(journalLines.entryId, journalEntries.id))
    .innerJoin(accounts, eq(journalLines.accountId, accounts.id))
    .where(
      and(
        eq(journalEntries.orgId, orgId),
        eq(journalEntries.status, "posted"),
        gte(journalEntries.entryDate, from),
        lte(journalEntries.entryDate, to),
        sql`${accounts.type} in ('revenue', 'expense')`,
      ),
    )
    .groupBy(accounts.id, accounts.code, accounts.name, accounts.type);

  let totalRevenue = 0;
  let totalExpense = 0;
  const items: Array<{ accountId: string; code: string; name: string; type: string; amount: number }> = [];

  for (const row of rows) {
    const debit = toNum(row.debit);
    const credit = toNum(row.credit);
    const amount =
      row.type === "revenue" ? credit - debit : debit - credit;
    if (row.type === "revenue") totalRevenue += amount;
    else totalExpense += amount;
    if (breakdown) {
      items.push({
        accountId: row.accountId,
        code: row.code,
        name: row.name,
        type: row.type,
        amount,
      });
    }
  }

  return {
    totalRevenue,
    totalExpense,
    netIncome: totalRevenue - totalExpense,
    ...(breakdown ? { breakdown: items } : {}),
  };
}

export async function getBalanceSheet(orgId: string, asOf?: string) {
  const date = asOf ?? new Date().toISOString().slice(0, 10);
  const balances = await accountBalances(orgId, date);

  const totalAssets = balances
    .filter((a) => a.type === "asset")
    .reduce((s, a) => s + a.balance, 0);
  const totalLiabilities = balances
    .filter((a) => a.type === "liability")
    .reduce((s, a) => s + a.balance, 0);
  const equityAccounts = balances
    .filter((a) => a.type === "equity")
    .reduce((s, a) => s + a.balance, 0);
  const netIncome =
    balances.filter((a) => a.type === "revenue").reduce((s, a) => s + a.balance, 0) -
    balances.filter((a) => a.type === "expense").reduce((s, a) => s + a.balance, 0);

  return {
    asOf: date,
    totalAssets,
    totalLiabilities,
    equityWithIncome: equityAccounts + netIncome,
  };
}

export async function getTrialBalance(orgId: string, asOf?: string) {
  const date = asOf ?? new Date().toISOString().slice(0, 10);
  const rows = await db
    .select({
      accountId: accounts.id,
      code: accounts.code,
      name: accounts.name,
      debit: sql<string>`coalesce(sum(${journalLines.debit}), 0)`,
      credit: sql<string>`coalesce(sum(${journalLines.credit}), 0)`,
    })
    .from(journalLines)
    .innerJoin(journalEntries, eq(journalLines.entryId, journalEntries.id))
    .innerJoin(accounts, eq(journalLines.accountId, accounts.id))
    .where(
      and(
        eq(journalEntries.orgId, orgId),
        eq(journalEntries.status, "posted"),
        lte(journalEntries.entryDate, date),
      ),
    )
    .groupBy(accounts.id, accounts.code, accounts.name);

  let totalDebit = 0;
  let totalCredit = 0;
  const accountsList = rows.map((row) => {
    const debit = toNum(row.debit);
    const credit = toNum(row.credit);
    totalDebit += debit;
    totalCredit += credit;
    return { accountId: row.accountId, code: row.code, name: row.name, debit, credit };
  });

  return { asOf: date, totalDebit, totalCredit, accounts: accountsList };
}

export async function getCashFlow(orgId: string, from: string, to: string) {
  const cashAccounts = await db
    .select({ id: accounts.id })
    .from(accounts)
    .where(and(eq(accounts.orgId, orgId), eq(accounts.isCash, true)));

  const cashIds = cashAccounts.map((a) => a.id);
  if (cashIds.length === 0) {
    return { from, to, cashIn: 0, cashOut: 0, netCashFlow: 0 };
  }

  const rows = await db
    .select({
      debit: sql<string>`coalesce(sum(${journalLines.debit}), 0)`,
      credit: sql<string>`coalesce(sum(${journalLines.credit}), 0)`,
    })
    .from(journalLines)
    .innerJoin(journalEntries, eq(journalLines.entryId, journalEntries.id))
    .where(
      and(
        eq(journalEntries.orgId, orgId),
        eq(journalEntries.status, "posted"),
        gte(journalEntries.entryDate, from),
        lte(journalEntries.entryDate, to),
        inArray(journalLines.accountId, cashIds),
      ),
    );

  const cashIn = toNum(rows[0]?.debit);
  const cashOut = toNum(rows[0]?.credit);
  return { from, to, cashIn, cashOut, netCashFlow: cashIn - cashOut };
}

File: backend-go/internal/reports/aggregates.go
package reports

import (
	"context"
	"math"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// DashboardMetrics matches the TS dashboard payload.
type DashboardMetrics struct {
	Cash            float64 `json:"cash"`
	PeriodRevenue   float64 `json:"periodRevenue"`
	PeriodNetIncome float64 `json:"periodNetIncome"`
	Receivables     float64 `json:"receivables"`
	Payables        float64 `json:"payables"`
	Equity          float64 `json:"equity"`
}

// DateRange is an optional period filter.
type DateRange struct {
	From string
	To   string
}

type accountBalance struct {
	AccountID string
	Code      string
	Name      string
	Type      string
	IsCash    bool
	Debit     float64
	Credit    float64
	Balance   float64
}

func toNum(v float64) float64 {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return 0
	}
	return v
}

func round2(n float64) float64 {
	return math.Round(n*100) / 100
}

func defaultPeriod() (from, to string) {
	now := time.Now().UTC()
	from = time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC).Format("2006-01-02")
	to = now.Format("2006-01-02")
	return from, to
}

func balanceForType(accountType string, debit, credit float64) float64 {
	if accountType == "asset" || accountType == "expense" {
		return debit - credit
	}
	return credit - debit
}

func accountBalances(ctx context.Context, db *pgxpool.Pool, orgID string, asOf string) ([]accountBalance, error) {
	query := `
		SELECT a.id, a.code, a.name, a.type, a.is_cash,
			coalesce(sum(jl.debit::numeric), 0),
			coalesce(sum(jl.credit::numeric), 0)
		FROM journal_lines jl
		INNER JOIN journal_entries je ON jl.entry_id = je.id
		INNER JOIN accounts a ON jl.account_id = a.id
		WHERE je.org_id = $1 AND je.status = 'posted'`
	args := []any{orgID}
	if asOf != "" {
		query += ` AND je.entry_date <= $2`
		args = append(args, asOf)
	}
	query += `
		GROUP BY a.id, a.code, a.name, a.type, a.is_cash`

	rows, err := db.Query(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []accountBalance
	for rows.Next() {
		var b accountBalance
		if err := rows.Scan(&b.AccountID, &b.Code, &b.Name, &b.Type, &b.IsCash, &b.Debit, &b.Credit); err != nil {
			return nil, err
		}
		b.Debit = toNum(b.Debit)
		b.Credit = toNum(b.Credit)
		b.Balance = balanceForType(b.Type, b.Debit, b.Credit)
		out = append(out, b)
	}
	return out, rows.Err()
}

func periodActivity(ctx context.Context, db *pgxpool.Pool, orgID, from, to string) (periodRevenue, periodExpense, periodNetIncome float64, err error) {
	rows, err := db.Query(ctx, `
		SELECT a.type,
			coalesce(sum(jl.debit::numeric), 0),
			coalesce(sum(jl.credit::numeric), 0)
		FROM journal_lines jl
		INNER JOIN journal_entries je ON jl.entry_id = je.id
		INNER JOIN accounts a ON jl.account_id = a.id
		WHERE je.org_id = $1 AND je.status = 'posted'
			AND je.entry_date >= $2 AND je.entry_date <= $3
		GROUP BY a.type
	`, orgID, from, to)
	if err != nil {
		return 0, 0, 0, err
	}
	defer rows.Close()

	for rows.Next() {
		var (
			typ            string
			debit, credit  float64
		)
		if err := rows.Scan(&typ, &debit, &credit); err != nil {
			return 0, 0, 0, err
		}
		debit, credit = toNum(debit), toNum(credit)
		switch typ {
		case "revenue":
			periodRevenue += credit - debit
		case "expense":
			periodExpense += debit - credit
		}
	}
	if err := rows.Err(); err != nil {
		return 0, 0, 0, err
	}
	return periodRevenue, periodExpense, periodRevenue - periodExpense, nil
}

// GetDashboardMetrics returns cash, P&L period metrics, AR/AP, and equity.
func GetDashboardMetrics(ctx context.Context, db *pgxpool.Pool, orgID string, range_ *DateRange) (*DashboardMetrics, error) {
	from, to := defaultPeriod()
	if range_ != nil && range_.From != "" && range_.To != "" {
		from, to = range_.From, range_.To
	}

	balances, err := accountBalances(ctx, db, orgID, to)
	if err != nil {
		return nil, err
	}
	periodRevenue, _, periodNetIncome, err := periodActivity(ctx, db, orgID, from, to)
	if err != nil {
		return nil, err
	}

	var cash float64
	for _, a := range balances {
		if a.IsCash {
			cash += a.Balance
		}
	}

	receivables, err := sumOpenItems(ctx, db, orgID, "receivable")
	if err != nil {
		return nil, err
	}
	payables, err := sumOpenItems(ctx, db, orgID, "payable")
	if err != nil {
		return nil, err
	}

	if receivables == 0 {
		for _, a := range balances {
			if a.Type == "asset" && strings.HasPrefix(a.Code, "13") {
				receivables += a.Balance
			}
		}
	}
	if payables == 0 {
		for _, a := range balances {
			if a.Type == "liability" && strings.HasPrefix(a.Code, "21") {
				payables += a.Balance
			}
		}
	}

	var equityAccounts, allTimeRevenue, allTimeExpense float64
	for _, a := range balances {
		switch a.Type {
		case "equity":
			equityAccounts += a.Balance
		case "revenue":
			allTimeRevenue += a.Balance
		case "expense":
			allTimeExpense += a.Balance
		}
	}

	return &DashboardMetrics{
		Cash:            cash,
		PeriodRevenue:   periodRevenue,
		PeriodNetIncome: periodNetIncome,
		Receivables:     receivables,
		Payables:        payables,
		Equity:          equityAccounts + (allTimeRevenue - allTimeExpense),
	}, nil
}

func sumOpenItems(ctx context.Context, db *pgxpool.Pool, orgID, itemType string) (float64, error) {
	var sum float64
	err := db.QueryRow(ctx, `
		SELECT coalesce(sum(balance_amount::numeric), 0)
		FROM open_items WHERE org_id = $1 AND type = $2
	`, orgID, itemType).Scan(&sum)
	return toNum(sum), err
}

// GetProfitLoss returns period P&L totals, optionally with account breakdown.
func GetProfitLoss(ctx context.Context, db *pgxpool.Pool, orgID, from, to string, breakdown bool) (map[string]any, error) {
	rows, err := db.Query(ctx, `
		SELECT a.id, a.code, a.name, a.type,
			coalesce(sum(jl.debit::numeric), 0),
			coalesce(sum(jl.credit::numeric), 0)
		FROM journal_lines jl
		INNER JOIN journal_entries je ON jl.entry_id = je.id
		INNER JOIN accounts a ON jl.account_id = a.id
		WHERE je.org_id = $1 AND je.status = 'posted'
			AND je.entry_date >= $2 AND je.entry_date <= $3
			AND a.type IN ('revenue', 'expense')
		GROUP BY a.id, a.code, a.name, a.type
	`, orgID, from, to)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var totalRevenue, totalExpense float64
	var items []map[string]any
	for rows.Next() {
		var (
			id, code, name, typ string
			debit, credit       float64
		)
		if err := rows.Scan(&id, &code, &name, &typ, &debit, &credit); err != nil {
			return nil, err
		}
		debit, credit = toNum(debit), toNum(credit)
		var amount float64
		if typ == "revenue" {
			amount = credit - debit
			totalRevenue += amount
		} else {
			amount = debit - credit
			totalExpense += amount
		}
		if breakdown {
			items = append(items, map[string]any{
				"accountId": id,
				"code":      code,
				"name":      name,
				"type":      typ,
				"amount":    amount,
			})
		}
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	out := map[string]any{
		"totalRevenue": totalRevenue,
		"totalExpense": totalExpense,
		"netIncome":    totalRevenue - totalExpense,
	}
	if breakdown {
		if items == nil {
			items = []map[string]any{}
		}
		out["breakdown"] = items
	}
	return out, nil
}

// GetBalanceSheet returns assets, liabilities, and equity including net income.
func GetBalanceSheet(ctx context.Context, db *pgxpool.Pool, orgID, asOf string) (map[string]any, error) {
	date := asOf
	if date == "" {
		date = time.Now().UTC().Format("2006-01-02")
	}
	balances, err := accountBalances(ctx, db, orgID, date)
	if err != nil {
		return nil, err
	}

	var totalAssets, totalLiabilities, equityAccounts, revenue, expense float64
	for _, a := range balances {
		switch a.Type {
		case "asset":
			totalAssets += a.Balance
		case "liability":
			totalLiabilities += a.Balance
		case "equity":
			equityAccounts += a.Balance
		case "revenue":
			revenue += a.Balance
		case "expense":
			expense += a.Balance
		}
	}

	return map[string]any{
		"asOf":             date,
		"totalAssets":      totalAssets,
		"totalLiabilities": totalLiabilities,
		"equityWithIncome": equityAccounts + (revenue - expense),
	}, nil
}

// GetTrialBalance returns debit/credit totals per account as of a date.
func GetTrialBalance(ctx context.Context, db *pgxpool.Pool, orgID, asOf string) (map[string]any, error) {
	date := asOf
	if date == "" {
		date = time.Now().UTC().Format("2006-01-02")
	}

	rows, err := db.Query(ctx, `
		SELECT a.id, a.code, a.name,
			coalesce(sum(jl.debit::numeric), 0),
			coalesce(sum(jl.credit::numeric), 0)
		FROM journal_lines jl
		INNER JOIN journal_entries je ON jl.entry_id = je.id
		INNER JOIN accounts a ON jl.account_id = a.id
		WHERE je.org_id = $1 AND je.status = 'posted' AND je.entry_date <= $2
		GROUP BY a.id, a.code, a.name
		ORDER BY a.code
	`, orgID, date)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var totalDebit, totalCredit float64
	accountsList := []map[string]any{}
	for rows.Next() {
		var (
			id, code, name  string
			debit, credit   float64
		)
		if err := rows.Scan(&id, &code, &name, &debit, &credit); err != nil {
			return nil, err
		}
		debit, credit = toNum(debit), toNum(credit)
		totalDebit += debit
		totalCredit += credit
		accountsList = append(accountsList, map[string]any{
			"accountId": id,
			"code":      code,
			"name":      name,
			"debit":     debit,
			"credit":    credit,
		})
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	return map[string]any{
		"asOf":        date,
		"totalDebit":  totalDebit,
		"totalCredit": totalCredit,
		"accounts":    accountsList,
	}, nil
}

// GetCashFlow returns cash in/out for the period from cash accounts.
func GetCashFlow(ctx context.Context, db *pgxpool.Pool, orgID, from, to string) (map[string]any, error) {
	rows, err := db.Query(ctx, `
		SELECT id FROM accounts WHERE org_id = $1 AND is_cash = true
	`, orgID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var cashIDs []string
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			return nil, err
		}
		cashIDs = append(cashIDs, id)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	if len(cashIDs) == 0 {
		return map[string]any{
			"from": from, "to": to,
			"cashIn": 0, "cashOut": 0, "netCashFlow": 0,
		}, nil
	}

	var debit, credit float64
	err = db.QueryRow(ctx, `
		SELECT coalesce(sum(jl.debit::numeric), 0), coalesce(sum(jl.credit::numeric), 0)
		FROM journal_lines jl
		INNER JOIN journal_entries je ON jl.entry_id = je.id
		WHERE je.org_id = $1 AND je.status = 'posted'
			AND je.entry_date >= $2 AND je.entry_date <= $3
			AND jl.account_id = ANY($4)
	`, orgID, from, to, cashIDs).Scan(&debit, &credit)
	if err != nil {
		return nil, err
	}
	cashIn, cashOut := toNum(debit), toNum(credit)
	return map[string]any{
		"from": from, "to": to,
		"cashIn": cashIn, "cashOut": cashOut, "netCashFlow": cashIn - cashOut,
	}, nil
}

// AgingBucket holds AR/AP aging totals.
type AgingBucket struct {
	Current    float64 `json:"current"`
	Days1to30  float64 `json:"days1to30"`
	Days31to60 float64 `json:"days31to60"`
	Days61to90 float64 `json:"days61to90"`
	Days90Plus float64 `json:"days90plus"`
	Total      float64 `json:"total"`
}

// GetAgingReport duplicates AR/AP aging SQL (avoids circular import with arap).
func GetAgingReport(ctx context.Context, db *pgxpool.Pool, orgID, kind string) (map[string]any, error) {
	if kind != "payable" {
		kind = "receivable"
	}
	today := time.Now().UTC().Truncate(24 * time.Hour)

	rows, err := db.Query(ctx, `
		SELECT oi.id, oi.contact_id, c.name, oi.description, oi.due_date, oi.balance_amount::numeric
		FROM open_items oi
		LEFT JOIN contacts c ON oi.contact_id = c.id
		WHERE oi.org_id = $1 AND oi.type = $2 AND oi.status IN ('open', 'partial')
	`, orgID, kind)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	buckets := AgingBucket{}
	items := []map[string]any{}

	for rows.Next() {
		var (
			id, description     string
			contactID           *string
			contactName         *string
			dueDate             *time.Time
			balance             float64
		)
		if err := rows.Scan(&id, &contactID, &contactName, &description, &dueDate, &balance); err != nil {
			return nil, err
		}
		balance = toNum(balance)
		if balance <= 0 {
			continue
		}

		bucket := "current"
		if dueDate != nil {
			due := time.Date(dueDate.Year(), dueDate.Month(), dueDate.Day(), 0, 0, 0, 0, time.UTC)
			diffDays := int(today.Sub(due).Hours() / 24)
			switch {
			case diffDays <= 0:
				bucket = "current"
			case diffDays <= 30:
				bucket = "days1to30"
			case diffDays <= 60:
				bucket = "days31to60"
			case diffDays <= 90:
				bucket = "days61to90"
			default:
				bucket = "days90plus"
			}
		}

		switch bucket {
		case "current":
			buckets.Current += balance
		case "days1to30":
			buckets.Days1to30 += balance
		case "days31to60":
			buckets.Days31to60 += balance
		case "days61to90":
			buckets.Days61to90 += balance
		case "days90plus":
			buckets.Days90Plus += balance
		}
		buckets.Total += balance

		var dueStr any
		if dueDate != nil {
			dueStr = dueDate.Format("2006-01-02")
		} else {
			dueStr = nil
		}
		items = append(items, map[string]any{
			"contactId":     contactID,
			"contactName":   contactName,
			"openItemId":    id,
			"description":   description,
			"dueDate":       dueStr,
			"balanceAmount": balance,
			"bucket":        bucket,
		})
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	buckets.Current = round2(buckets.Current)
	buckets.Days1to30 = round2(buckets.Days1to30)
	buckets.Days31to60 = round2(buckets.Days31to60)
	buckets.Days61to90 = round2(buckets.Days61to90)
	buckets.Days90Plus = round2(buckets.Days90Plus)
	buckets.Total = round2(buckets.Total)

	return map[string]any{
		"kind":    kind,
		"buckets": buckets,
		"items":   items,
	}, nil
}
